scanner/src/main/kotlin/scanners/fossid/FossId.kt

```diff
@@ -0,0 +1,396 @@
+     */
+    private suspend fun createScan(projectCode: String, url: String, revision: String): String {
+        val formatter = DateTimeFormatter.ofPattern("'$projectCode'_yyyyMMdd_HHmmss")
+        val scanCode = formatter.format(LocalDateTime.now())
+        val response = service.createScan(user, apiKey, projectCode, scanCode, url, revision)
+            .checkResponse("create scan")
+        val scanId = response.data?.get("scan_id")
+        requireNotNull(scanId) {
+            "Scan was not triggered..."
+        }
+        log.info { "Scan has been created with id=$scanId." }
+        return scanCode
+    }
+
+    /**
+     * Check a scan: the repository must have been downloaded and the scan must have completed.
+     * This function triggers the scan if needed.
+     */
+    private suspend fun checkScan(scanCode: String) {
+        waitDownloadComplete(scanCode)
+        val response = service.checkScanStatus(user, apiKey, scanCode)
+            .checkResponse("check scan status", false)
+        if (response.data?.state == ScanState.NOT_STARTED) {
+            service.runScan(user, apiKey, scanCode)
+                .checkResponse("trigger scan", false)
+            log.info { "Scan has been launched." }
+
+            waitScanComplete(scanCode)
+        }
+    }
+
+    /**
+     * Wait for the lambda [waitLoop] to return true, waiting [delay] between each invocation.
+     * [waitLoop] should return true if the wait must be interrupted, false otherwise.
+     *
+     * A [timeout] will be honored and null will be returned if the timeout has been reached.
+     */
+    private suspend fun wait(timeout: Long, delay: Long, waitLoop: suspend () -> Boolean) =
+        withTimeoutOrNull(timeout) {
+            var isFinished = false
+            while (!isFinished) {
+                isFinished = waitLoop()
+
+                if (!isFinished) {
+                    delay(delay)
+                }
+            }
+        }
+
+    /**
+     * Wait until the repository of a scan with [scanCode] has been downloaded.
+     */
+    private suspend fun waitDownloadComplete(scanCode: String) {
+        val timeout = wait(WAIT_INTERVAL_MS * WAIT_REPETITION, WAIT_INTERVAL_MS) {
+            this@FossId.log.info { "Check download status for '$scanCode'." }
+
+            val response = service.checkDownloadStatus(user, apiKey, scanCode)
+                .checkResponse("check download status")
+
+            if (response.data == DownloadStatus.FINISHED) {
+                true
+            } else {
+                /*
+        There is a weird bug with the FossId server with version < 20.2: sometimes the download is
+        complete but it stays in state "NOT FINISHED". Therefore we check the output of Git Fetch to
+        find out if the download is indeed done
+                 */
+                val message = response.message
+                if (message == null || !GIT_FETCH_DONE_PATTERN.matcher(message).find()) {
+                    false
```

Comment:
In general, the above code could be rewritten as

            if (response.data == DownloadStatus.FINISHED) return@wait true

            val message = response.message
            if (message == null || !GIT_FETCH_DONE_PATTERN.matcher(message).matches()) return@wait false

to reduce nesting.